slurm_pipeline/sacct.py

```diff
@@ -0,0 +1,106 @@
+from os import getlogin
+import subprocess
+
+from error import SAcctError
+from datetime import datetime, timedelta
+
+class SAcct(object):
+    """
+    Fetch information about job id status from sacct.
+
+    @param sacctArgs: A C{list} of C{str} arguments to pass to sacct
+        (including the 'sacct' command itself). If C{None}, the user's
+        login name will be appended to sacct -u.
+    """
+    def __init__(self, sacctArgs=None):
+        starttime = (datetime.now() - timedelta(days=7)).strftime("%Y-%m-%d")
+        args = sacctArgs or ['sacct', '-p', '-l', '-u', getlogin(), '-S', starttime]
+        try:
+            out = subprocess.check_output(args, universal_newlines=True)
+        except OSError as e:
+            raise SAcctError("Encountered OSError (%s) when running '%s'" %
+                              (e, ' '.join(args)))
+
+        columnInfo = {
+            'JobID': {
+                'attr': 'jobid',
+                'offset': None,
+            },
+            'Elapsed': {
+                'attr': 'elapsed',
+                'offset': None,
+            },
+            'State': {
+                'attr': 'state',
+                'offset': None,
+            }
+        }
+
+        jobs = {}
+
+        for count, line in enumerate(out.split('\n')):
+            fields = line.split('|')
+            if count == 0:
+                # Process header line.
+                colsFound = 0
+                for offset, title in enumerate(fields):
+                    if title in columnInfo:
+                        if columnInfo[title]['offset'] is not None:
+                            raise SAcctError(
+                                "Multiple columns with title %r found in '%s' "
+                                'output' % (title, ' '.join(args)))
+                        else:
+                            columnInfo[title]['offset'] = offset
+                            colsFound += 1
+                if colsFound != len(columnInfo):
+                    notFound = [title for title in columnInfo
+                                if columnInfo[title]['offset'] is None]
+                    raise SAcctError(
+                        "Required column%s (%s) not found in '%s' "
+                        'output' % ('' if len(notFound) == 1 else 's',
+                                    ', '.join(sorted(notFound)),
+                                    ' '.join(args)))
+            elif count == 1:
```

Comment:
What version of SLURM do you need this for? In the version I have access to (17.02.10) there is only one header line.

Reply:
```
$ sinfo -o "%v"
VERSION
17.02
```

But when I simply do 
```
$ sacct
       JobID    JobName  Partition    Account  AllocCPUS      State ExitCode 
------------ ---------- ---------- ---------- ---------- ---------- --------
```

If I do
$ sacct -p -l -u sschmitt
JobID|JobIDRaw|JobName|Partition|MaxVMSize|MaxVMSizeNode|MaxVMSizeTask|AveVMSize|MaxRSS|MaxRSSNode|MaxRSSTask|AveRSS|MaxPages|MaxPagesNode|MaxPagesTask|AvePages|MinCPU|MinCPUNode|MinCPUTask|AveCPU|NTasks|AllocCPUS|Elapsed|State|ExitCode|AveCPUFreq|ReqCPUFreqMin|ReqCPUFreqMax|ReqCPUFreqGov|ReqMem|ConsumedEnergy|MaxDiskRead|MaxDiskReadNode|MaxDiskReadTask|AveDiskRead|MaxDiskWrite|MaxDiskWriteNode|MaxDiskWriteTask|AveDiskWrite|AllocGRES|ReqGRES|ReqTRES|AllocTRES|
1116681|1116681|bash|short|310860K|HBPHost6|0|29292K|3260K|HBPHost6|0|2224K|3K|HBPHost6|0|3K|00:00:00|HBPHost6|0|00:00:00|1|2|00:01:31|COMPLETED|0:0|1.10G|Unknown|Unknown|Unknown|3750Mc|0|0.84M|HBPHost6|0|0.84M|0.01M|HBPHost6|0|0.01M|||cpu=1,mem=3750M,node=1|cpu=2,mem=7500M,node=1|
```
I also get no separation line... It's probably better to do the checking explicitly

Reply:
Huh, that's a bit weird. I'll do the checking, as you suggest. Thanks.